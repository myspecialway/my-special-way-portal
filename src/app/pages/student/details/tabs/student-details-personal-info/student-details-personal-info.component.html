<div class="msw-student-body-ph">
  <div class="msw-student-header-nav-name" *ngIf="!isNewStudent && student">
    {{ student.firstname }} {{ student.lastname }}
  </div>
  <div class="msw-student-body">
    <div class="msw-student-user-pic-ph">
      <img src="http://via.placeholder.com/140x100">
    </div>
    <div class="msw-student-user-form">
      <form class="msw-student-details-form" #studentDetailsForm="ngForm" *ngIf="student">

        <div class="msw-student-main-details">
          <mat-form-field class="firstname">
            <input matInput placeholder="שם פרטי" [(ngModel)]="student.firstname" name="firstname" #firstName="ngModel" required>
            <mat-error data-test-id="firstname-err" *ngIf="firstName.invalid">שם פרטי הינו שדה חובה</mat-error>
          </mat-form-field>
          <mat-form-field class="lastname">
            <input matInput placeholder="שם משפחה" [(ngModel)]="student.lastname" name="lastname" #lastName="ngModel" required>
            <mat-error data-test-id="lastname-err" *ngIf="lastName.invalid">שם משפחה הינו שדה חובה</mat-error>

          </mat-form-field>
          <div class="msw-student-details-gender-switch">
            <div class="msw-student-details-gender">מין</div>
            <mat-button-toggle-group name="fontStyle" aria-label="Font Style" [(ngModel)]="student.gender" name="gender" required>
              <mat-button-toggle value="FEMALE">
                <i class="material-icons"> face </i>
              </mat-button-toggle>
              <mat-button-toggle value="MALE">
                <i class="material-icons"> face </i>
              </mat-button-toggle>
            </mat-button-toggle-group>
          </div>
        </div>

        <mat-form-field class="classname">
          <mat-select placeholder="כיתה" [(ngModel)]="student.class._id" #classId="ngModel" name="class_id" required>
            <mat-option *ngFor="let classname of classes" [value]="classname._id">
              {{classname.name}}
            </mat-option>
          </mat-select>
          <mat-error data-test-id="class-id-err" *ngIf="classId.hasError('required')">כיתה הינו שדה חובה</mat-error>
        </mat-form-field>

        <mat-form-field class="username">
          <input matInput placeholder="שם משתמש" [(ngModel)]="student.username" name="username" #username="ngModel"
                 autocomplete="new-username" required [minlength]="5" [pattern]="'^[A-Za-z]+$'">
          <mat-error data-test-id="username-err" *ngIf="username.hasError('required')">שם משתמש הינו שדה חובה</mat-error>
          <mat-error data-test-id="username-format-err" *ngIf="username.hasError('minlength') || username.hasError('pattern')">שם משתמש חייב להיות לפחות חמישה תווים באנגלית</mat-error>
        </mat-form-field>
        <mat-form-field class="password">
          <input matInput placeholder="סיסמא" type="password" [(ngModel)]="student.password" #password="ngModel" name="password" autocomplete="new-password" required>
          <mat-error data-test-id="password-err" *ngIf="!password.valid" class="help-block">נא הכנס סיסמה</mat-error>
        </mat-form-field>

        <div mat-dialog-actions>
          <button mat-button data-test-id="save-button" class="msw-button-add" (click)="isNewStudent ? addStudent() : updateStudent(studentDetailsForm)"
                  [disabled]="!studentDetailsForm.valid" >שמור</button>
        </div>
      </form>
    </div>
  </div>
</div>
