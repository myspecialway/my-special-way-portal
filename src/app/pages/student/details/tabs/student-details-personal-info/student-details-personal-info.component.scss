:host {
  height: 100%;
}

button {
  background-color: #6a00cf;
  border-radius: 3px;
  color: #fff;
  height: 32px;
  width: 120px;
}

.msw-student-main-details {
  display: flex;
  align-items: center;
  justify-content: space-between;

  .firstname,
  .lastname {
    margin-left: 24px;
  }
}

.msw-student {
  height: 100%;
  background-color: #fafafa;
  padding-top: 100px;
}

.msw-student-header-nav-name {
  font-size: 32px;
  padding-top: 5px;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
  position: absolute;
  top: 125px;
  right: 38px;
}

.msw-student-body-ph {
  display: flex;
  align-items: center;
  justify-content: center;
  padding-top: 30px;
}

.msw-student-body {
  height: 475px;
  border-radius: 3px;
  background-color: #ffffff;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.08);
  display: flex;
  align-items: center;
  flex-direction: column;
  padding: 48px;
}

.msw-student-details-gender {
  font-size: 12px;
  line-height: 1.42857;
  color: #aaaaaa;
  font-weight: 400;
  text-align: right;
}

.msw-button-saved-indicator {
  background-color: #2ea21e;
}

.mat-raised-button {
  line-height: inherit;
}

.msw-student-user-pic-ph {
}

.msw-student-user-form {
}
