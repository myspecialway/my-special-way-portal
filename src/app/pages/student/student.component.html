<div data-test-id="students-page" class="main-content msw-student-container">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">תלמידים</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button data-test-id="add-new-student-button" mat-raised-button color="primary" routerLink="/student/_new_"  *ngIf="mayAddStudent">
          <i class="material-icons">
            add
          </i>
          הוסף תלמיד חדש
        </button>
      </div>
    </div>
    <div class="row">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table [dataSource]="dataSource" class="msw-table" matSort matSortDisableClear>
          <ng-container matColumnDef="studentName">
            <mat-header-cell *matHeaderCellDef>שם</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{ row.firstname }} {{ row.lastname }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="gradeId">
            <mat-header-cell *matHeaderCellDef>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" [routerLink]="['/student', row._id]">{{(row.class && row.class.name) || '' }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header>שם משתמש</mat-header-cell>
            <mat-cell *matCellDef="let row" class="username" [routerLink]="['/student', row._id]">{{ row.username }}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="deleteUser">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button mat-icon-button color="primary" [attr.data-test-id]="'delete-user-button-' + row.username" [disabled]="!mayDeleteStudent"
                      (click)="deleteStudent(row._id, row.firstname, row.lastname, row.gradeId, row.gender)"><i class="material-icons"> delete </i> </button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
