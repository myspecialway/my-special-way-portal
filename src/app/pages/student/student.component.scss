@import '../../../assets/scss/msw-table';

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}
.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.msw-student-container .mat-cell:last-child,
.msw-student-container .mat-header-cell:last-child {
  flex-grow: 0;
  flex-basis: 100px;
}
.msw-student-container .mat-header-cell {
  font-weight: bold;
}

.mat-column-deleteUser button {
  display: none;
}
.mat-row:hover {
  .mat-cell {
    color: #8c37cd;
  }
  .mat-column-deleteUser button {
    display: block;
  }
}
