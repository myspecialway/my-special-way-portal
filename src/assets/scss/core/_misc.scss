body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
