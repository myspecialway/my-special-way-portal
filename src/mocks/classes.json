[
  {
    "_id": "5b0116f8c1d99ef1b8e73245",
    "index": 0,
    "grade": "A",
    "number": 1,
    "description": "Tiltan",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f870a67d3a47e9f235",
    "index": 1,
    "grade": "A",
    "number": 2,
    "description": "Tapuz",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f82074861f963642f3",
    "index": 2,
    "grade": "A",
    "number": 3,
    "description": "Tapuch",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8b9e91d54d315372c",
    "index": 3,
    "grade": "A",
    "number": 4,
    "description": "Tut",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8895af89de5563236",
    "index": 4,
    "grade": "B",
    "number": 1,
    "description": "Dafna",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f870788b1e9f99b814",
    "index": 5,
    "grade": "B",
    "number": 2,
    "description": "Dekel",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f896b61b3a5c135a1d",
    "index": 6,
    "grade": "B",
    "number": 3,
    "description": "Duvdevan",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f855ecf6fd19ae77fa",
    "index": 7,
    "grade": "B",
    "number": 4,
    "description": "Dolev",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 1,
    "description": "Egoz",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 2,
    "description": "Agas",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 3,
    "description": "Erez",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 4,
    "description": "Oren",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 5,
    "description": "Alon",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f87554b3534cfde9b7",
    "index": 9,
    "grade": "D",
    "number": 1,
    "description": "Shaked",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8cc498da37f361b2f",
    "index": 10,
    "grade": "D",
    "number": 2,
    "description": "Gefen",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f832a6de313076c3b1",
    "index": 11,
    "grade": "D",
    "number": 3,
    "description": "Shezif",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f832a6de313076c3b1",
    "index": 11,
    "grade": "D",
    "number": 4,
    "description": "Shesek",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f841bc308bff51f0b8",
    "index": 12,
    "grade": "E",
    "number": 1,
    "description": "Rimon",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8897ec61406fc591f",
    "index": 13,
    "grade": "E",
    "number": 2,
    "description": "Teena",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f867a266b544055fc3",
    "index": 14,
    "grade": "E",
    "number": 3,
    "description": "Zait",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f846bd827beedec579",
    "index": 15,
    "grade": "F",
    "number": 1,
    "description": "Banana",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8ab446a3271359472",
    "index": 16,
    "grade": "F",
    "number": 2,
    "description": "Balon",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8f0319d311adbd5b8",
    "index": 17,
    "grade": "F",
    "number": 3,
    "description": "Bait",
    "num_of_students": 20
  }
]
