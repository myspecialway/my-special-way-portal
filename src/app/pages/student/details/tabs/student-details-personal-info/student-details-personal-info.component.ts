import { Component, OnInit } from '@angular/core';
import { StudentService } from '../../../services/student.service';
import Student, { Gender } from '../../../../../models/student.model';
import { ClassService } from '../../../../class/services/class.graphql.service';
import { Class } from '../../../../../models/class.model';
import { ActivatedRoute, Router } from '@angular/router';
import { SubscriptionCleaner } from '../../../../../decorators/SubscriptionCleaner.decorator';

@Component({
  selector: 'app-student-details-personal-info',
  templateUrl: './student-details-personal-info.component.html',
  styleUrls: ['./student-details-personal-info.component.scss'],
})
export class StudentDetailsPersonalInfoComponent implements OnInit {
  student: Student;
  classes: Class[];
  isNewStudent: boolean;
  idOrNew: string;

  @SubscriptionCleaner()
  subCollector;

  constructor(
    private studentService: StudentService,
    private classService: ClassService,
    private router: Router,
    private route: ActivatedRoute,
  ) {}

  ngOnInit(): void {
    if (this.route && this.route.parent) {
      this.subCollector.add(
        this.route.parent.params.subscribe((params) => {
          this.idOrNew = params.idOrNew;
        }),
      );
    }

    this.populateClasses();
    this.populateStudent();
  }

  async populateStudent() {
    if (this.idOrNew === '_new_') {
      this.isNewStudent = true;
      this.student = this.createNewStudent();
    } else {
      this.isNewStudent = false;
      try {
        this.student = await this.studentService.getById(this.idOrNew);
      } catch (error) {
        // TODO: implement error handling on UI
        console.error('Error handling not implemented');
        throw error;
      }
    }
  }

  createNewStudent(): Student {
    const student: Student = new Student();
    student.firstname = '';
    student.lastname = '';
    student.username = '';
    student.password = '';
    student.gender = Gender.FEMALE;
    student.class = new Class();
    return student;
  }

  async populateClasses() {
    try {
      this.classService.getAllClasses().subscribe((classes) => {
        this.classes = [...classes];
      });
    } catch (error) {
      // TODO: implement error handling on UI
      console.error('Error handling not implemented');
      throw error;
    }
  }

  async addStudent() {
    try {
      await this.studentService.create(this.student);
      this.router.navigate(['/student']);
    } catch (error) {
      // TODO: implement error handling on UI
      console.error('Error handling not implemented');
      throw error;
    }
  }

  async updateStudent(student) {
    student.form.value._id = this.student._id;
    try {
      await this.studentService.update(student.form.value);
      this.router.navigate(['/student']);
    } catch (error) {
      // TODO: implement error handling on UI
      console.error('Error handling not implemented');
      throw error;
    }
  }
}
