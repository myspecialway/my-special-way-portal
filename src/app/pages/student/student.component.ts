import { first } from 'rxjs/operators';
import { Component, OnInit, ViewChild, ElementRef } from '@angular/core';
import { MatTableDataSource, MatSort, MatDialog } from '@angular/material';
import { DeleteStudentDialogComponent } from './dialogs/delete/delete-student.dialog';
import { StudentService } from './services/student.service';
import Student from '../../models/student.model';
import { SubscriptionCleaner } from '../../decorators/SubscriptionCleaner.decorator';
@Component({
  selector: 'app-student',
  templateUrl: './student.component.html',
  styleUrls: ['./student.component.scss'],
})
export class StudentComponent implements OnInit {
  displayedColumns = ['studentName', 'gradeId', 'username', 'deleteUser'];
  dataSource = new MatTableDataSource<Student>();

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  @SubscriptionCleaner()
  subCollector;

  constructor(private studentService: StudentService, private dialog: MatDialog) {}

  ngOnInit() {
    this.dataSource.sort = this.sort;

    this.subCollector.add(
      this.studentService.getAllStudents().subscribe((data) => {
        this.dataSource.data = [...data];
      }),
    );
  }

  async deleteStudent(id: number, firstName: string, lastName: string, gradeId: string) {
    const dialogRef = this.dialog.open(DeleteStudentDialogComponent, {
      data: { id, firstName, lastName, gradeId },
    });

    dialogRef
      .afterClosed()
      .pipe(first())
      .subscribe(async (deletionConfirmed) => {
        if (!deletionConfirmed) {
          return;
        }

        try {
          await this.studentService.delete(id);
        } catch (error) {
          // TODO: implement error handling on UI
          console.error('Error handling not implemented');
          throw error;
        }
      });
  }
}
