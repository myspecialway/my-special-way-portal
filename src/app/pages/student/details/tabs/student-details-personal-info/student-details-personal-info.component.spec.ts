import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { MatHeaderRowDef, MatRowDef, MatHeaderRow, MatDialog, MatSort } from '@angular/material';
import { StudentDetailsComponent } from '../../student-details.component';
import { StudentService } from '../../../services/student.service';
import {
  Overlay,
  ScrollStrategyOptions,
  ScrollDispatcher,
  ViewportRuler,
  OverlayContainer,
  OverlayPositionBuilder,
  OverlayKeyboardDispatcher,
} from '@angular/cdk/overlay';
import { Platform } from '@angular/cdk/platform';
import { classTestData } from '../../../../../../mocks/assets/classes.mock';
import { Observable } from 'rxjs-compat';
import { FormsModule } from '@angular/forms';
import { ClassService } from '../../../../class/services/class.graphql.service';
import { ActivatedRoute, Router, RouterModule, Routes } from '@angular/router';
import { oneStudentTestData } from '../../../../../../mocks/assets/student.mock';
import { StudentDetailsPersonalInfoComponent } from './student-details-personal-info.component';

describe('Student Details Personal Info Component', () => {
  let studentServiceMock: Partial<StudentService>;
  let classServiceMock: Partial<ClassService>;
  let activatedRouteMock;
  let studentDialogMock: Partial<MatDialog>;
  let routerModuleMock: Partial<RouterModule>;

  beforeEach(async () => {
    studentServiceMock = {
      getAllStudents: jest.fn(),
      getById: jest.fn(),
      delete: jest.fn(),
      create: jest.fn(),
      update: jest.fn(),
    };

    routerModuleMock = {
      forRoot: jest.fn(),
      forChild: jest.fn(),
      navigate: jest.fn(),
    };

    classServiceMock = {
      getAllClasses: jest.fn().mockReturnValueOnce(Observable.of(classTestData.classes)),
    };

    studentDialogMock = {
      open: jest.fn().mockReturnValue({
        afterClosed: jest.fn().mockReturnValue(Observable.of(true)),
      }),
    };

    const routes: Routes = [
      { path: '', redirectTo: 'student', pathMatch: 'full' },
      {
        path: '',
        children: [
          {
            path: 'student/:idOrNew',
            component: StudentDetailsComponent,
            children: [
              { path: 'personalInfo', component: StudentDetailsPersonalInfoComponent },
              { path: '', component: StudentDetailsPersonalInfoComponent },
            ],
          },
        ],
      },
    ];

    TestBed.configureTestingModule({
      imports: [FormsModule, RouterModule.forRoot(routes)],
      declarations: [
        StudentDetailsComponent,
        StudentDetailsPersonalInfoComponent,
        MatHeaderRow,
        MatRowDef,
        MatHeaderRowDef,
        MatSort,
      ],
      providers: [
        StudentService,
        ClassService,
        { provide: MatDialog, useValue: studentDialogMock },
        { provide: StudentService, useValue: studentServiceMock },
        { provide: ClassService, useValue: classServiceMock },
        { provide: Router, useValue: routerModuleMock },
        Overlay,
        ScrollStrategyOptions,
        ScrollDispatcher,
        Platform,
        ViewportRuler,
        OverlayContainer,
        OverlayPositionBuilder,
        OverlayKeyboardDispatcher,
      ],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  describe('with _new_ student path', () => {
    beforeEach(async () => {
      TestBed.configureTestingModule({
        providers: [
          {
            provide: ActivatedRoute,
            useValue: {
              parent: {
                params: Observable.of({ idOrNew: '_new_' }),
              },
            },
          },
        ],
      });
    });

    it('should render the component as described in snapshot', () => {
      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      expect(fixture).toMatchSnapshot();
    });

    it('should load classes from service on page load ', async () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      expect(fixture.componentInstance.classes.length).toEqual(23);
    });

    it('should populate the isNewStudent to true when coming from path student/_new_', () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      expect(fixture.componentInstance.isNewStudent).toBe(true);
    });

    it('should call the create function on create student', async () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.create as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(1);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      fixture.componentInstance.addStudent();
      expect(studentServiceMock.create).toHaveBeenCalled();
    });

    it('should show indication that the changes were saved', async () => {
      jest.useFakeTimers();
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.create as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(1);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      expect(fixture.componentInstance.changesWereSaved).toBeFalsy();
      await fixture.componentInstance.addStudent();
      expect(fixture.componentInstance.changesWereSaved).toBeTruthy();
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 4000);
      jest.advanceTimersByTime(4000);
      expect(fixture.componentInstance.changesWereSaved).toBeFalsy();
    });

    it('should show indication when error happened on save changes', async () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.create as jest.Mock).mockImplementationOnce(() => {
        throw new Error('Mock Error');
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      expect(fixture.componentInstance.saveFailed).toBe(false);
      await fixture.componentInstance.addStudent();
      expect(fixture.componentInstance.saveFailed).toBe(true);
    });
  });

  describe('with edit student path', () => {
    beforeEach(async () => {
      activatedRouteMock = {
        params: {
          value: {
            id: '66',
          },
        },
      };

      TestBed.configureTestingModule({
        providers: [{ provide: ActivatedRoute, useValue: activatedRouteMock }],
      });
    });

    it('should load the student details from service on page load ', async () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.getById as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(oneStudentTestData.student);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      expect(fixture.componentInstance.student.firstname).toEqual('John1');
    });

    it('should populate the isNewStudent to false when coming from path student/:id', () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.getById as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(oneStudentTestData.student);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      expect(fixture.componentInstance.isNewStudent).toBe(false);
    });

    it('should call the update function on update student', async () => {
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.getById as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(oneStudentTestData.student);
      });

      (studentServiceMock.update as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(1);
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      fixture.componentInstance.updateStudent({ form: { value: { _id: '66' } } });
      expect(studentServiceMock.update).toHaveBeenCalled();
    });

    it('should show indication that the changes were saved', async () => {
      jest.useFakeTimers();
      (classServiceMock.getAllClasses as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(classTestData.classes);
      });

      (studentServiceMock.getById as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(oneStudentTestData.student);
      });

      (studentServiceMock.update as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(1);
      });
      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      expect(fixture.componentInstance.changesWereSaved).toBeFalsy();
      await fixture.componentInstance.updateStudent({ form: { value: { _id: '66' } } });
      expect(fixture.componentInstance.changesWereSaved).toBeTruthy();
      expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 4000);
    });

    it('should show indication when error happened on save changes', async () => {
      (studentServiceMock.getById as jest.Mock).mockImplementationOnce(() => {
        return Promise.resolve(oneStudentTestData.student);
      });

      (studentServiceMock.update as jest.Mock).mockImplementationOnce(() => {
        throw new Error('Mock Error');
      });

      const fixture = TestBed.createComponent(StudentDetailsPersonalInfoComponent);
      fixture.detectChanges();
      await fixture.whenRenderingDone();
      expect(fixture.componentInstance.saveFailed).toBe(false);
      await fixture.componentInstance.updateStudent({ form: { value: { _id: '66' } } });
      expect(fixture.componentInstance.saveFailed).toBe(true);
    });
  });
});
