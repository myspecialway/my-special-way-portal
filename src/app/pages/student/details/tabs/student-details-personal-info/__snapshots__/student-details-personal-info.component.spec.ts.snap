// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Student Details Personal Info Component with _new_ student path should render the component as described in snapshot 1`] = `
<app-student-details-personal-info
  classService={[Function Object]}
  route={[Function Object]}
  router={[Function Object]}
  studentService={[Function Object]}
>
  <div
    class="msw-student-body-ph"
  >
    
    <div
      class="msw-student-body"
    >
      <div
        class="msw-student-user-pic-ph"
      >
        <img
          src="http://via.placeholder.com/140x100"
        />
      </div>
      <div
        class="msw-student-user-form"
      >
        
      </div>
    </div>
  </div>
</app-student-details-personal-info>
`;
